### PR TITLE
request/response/error interceptors

### PR DESCRIPTION
**Pull request by artem-v**
**Motivation**

Give ability to intercept requests/response/exceptions  and supply client provided behavior at certain stages: on-request, on-response and on-exception  (exception information must come along with request which caused it)

**Updates**
Added exception/request/reponse interceptors at three levels: on ServiceCall, on ServiceInfo, and default ones on Microservices

### CODE REVIEW

**Comment by ronenhamias** [comment]
please describe this feature 
the motivation
how user is going to use it?
documentation and examples
please describe the top level api.

File: services-api/src/main/java/io/scalecube/services/methods/ServiceMethodInvoker.java [hunk 1]
```diff
@@ -190,10 +228,17 @@ private Object toRequest(ServiceMessage message) {
...
-        : ServiceMessage.builder().qualifier(methodInfo.qualifier()).data(response).build();
+  private ServiceMessage toResponse(Object response, Map<String, String> headers) {
+    if (response instanceof ServiceMessage) {
+      // return as is; it's assumed headers are copied inside service method
+      return (ServiceMessage) response;
+    }
+    // wrap response object into ServiceMessage and keep request headers
+    return ServiceMessage.builder()
+        .qualifier(methodInfo.qualifier())
+        .headers(headers)
```
Comment (segabriel) [inline services-api/src/main/java/io/scalecube/services/methods/ServiceMethodInvoker.java:268]: Cool, so didn't we set headers before?

Comment (artem-v) [reply services-api/src/main/java/io/scalecube/services/methods/ServiceMethodInvoker.java:268]: We didn't

File: services-api/src/main/java/io/scalecube/services/ServiceInfo.java [hunk 2]
```diff
@@ -47,12 +60,45 @@ public Authenticator authenticator() {
...
+
     private final Object serviceInstance;
     private final Map<String, String> tags = new HashMap<>();
+
     private ServiceProviderErrorMapper errorMapper;
     private ServiceMessageDataDecoder dataDecoder;
     private Authenticator authenticator;
+    private UnaryOperator<ServiceMessage> requestMapper;
+    private UnaryOperator<ServiceMessage> responseMapper;
+    private BiConsumer<ServiceMessage, Throwable> errorHandler;
```
Comment (ronenhamias) [inline services-api/src/main/java/io/scalecube/services/ServiceInfo.java:85]: should this be named errorMapper?

**Review by dmytro-lazebnyi — APPROVED** [review]

**Comment by ronenhamias** [comment]
can you show in example how i can trace only one specific method?

**Comment by artem-v** [comment]
Closing.

**Pull request closed without merging**